lib/semian.rb

```diff
@@ -150,10 +152,12 @@ def to_s
   #
   # Returns the registered resource.
   def register(name, **options)
-    circuit_breaker = create_circuit_breaker(name, **options)
-    bulkhead = create_bulkhead(name, **options)
-
-    resources[name] = ProtectedResource.new(bulkhead, circuit_breaker)
+    @lock.synchronize do
+      return resources[name] if resources[name]
```

Comment:
If you acquire the lock the first time, you'd set it—any other thread that'd race in here (because of the `||` clause in `retrieve_or_register`) would then receive the first resource created when it got the lock (but we want to wait for the first one that races here to return it).

Maybe I'm not understanding you correctly, but I am fairly sure this doesn't break any functionality?